lisa/analysis/tasks.py

```diff
@@ -228,47 +229,106 @@ def df_task_states(self, task, stringify=False):
             wkn_df = self.trace.df_events('sched_wakeup_new')
             wk_df = pd.concat([wk_df, wkn_df]).sort_index()
 
-        task_wakeup = wk_df[wk_df.pid == pid][['target_cpu', '__cpu']]
-        task_wakeup['curr_state'] = TaskState.TASK_WAKING
+        wk_df = wk_df[wk_df.success == 1][["pid", "target_cpu", "__cpu"]]
+        wk_df["curr_state"] = TaskState.TASK_WAKING
 
-        task_switches_df = sw_df[
-            (sw_df.prev_pid == pid) |
-            (sw_df.next_pid == pid)
-        ][['__cpu', 'prev_pid', 'prev_state']]
+        prev_sw_df = sw_df[["__cpu", "prev_pid", "prev_state"]].copy()
+        next_sw_df = sw_df[["__cpu", "next_pid"]].copy()
 
-        # This is a switch-in event
-        # (we don't care about the status of a task we are replacing)
-        task_switches_df.prev_state[task_switches_df.prev_pid != pid] = TaskState.TASK_ACTIVE
+        prev_sw_df.rename(columns={"prev_pid" : "pid", "prev_state" : "curr_state"},
+                          inplace=True)
 
-        task_switches_df = task_switches_df.drop(columns=["prev_pid"])
+        next_sw_df["curr_state"] = TaskState.TASK_ACTIVE
+        next_sw_df.rename(columns={'next_pid' : 'pid'}, inplace=True)
 
-        task_switches_df.rename(columns={'prev_state' : 'curr_state'}, inplace=True)
+        all_sw_df = prev_sw_df.append(next_sw_df, sort=False)
+        all_sw_df.sort_index(inplace=True)
 
         # Integer values are prefered here, otherwise the whole column
         # is converted to float64
-        task_switches_df['target_cpu'] = -1
+        all_sw_df['target_cpu'] = -1
+
+        df = all_sw_df.append(wk_df, sort=False)
+        df.sort_index(inplace=True)
```

Comment:
cannot we drop the sort at L245 and do just that once we already appended `next_`'s and `wk_`'s content to all_sw_df ?

Reply:
Did you try to use mergesort instead of quicksort?... it's sometimes more cache efficient and perhaps it can buy you something especially on small DFs or on very big ones not fitting in memory.

Reply:
> cannot we drop the sort at L245 and do just that once we already appended `next_`'s and `wk_`'s content to all_sw_df ?

Right, that should work.

> Did you try to use mergesort instead of quicksort?... it's sometimes more cache efficient and perhaps it can buy you something especially on small DFs or on very big ones not fitting in memory.

I just did and didn't notice any notable difference.